### DISCUSSION TITLE
Adding new column

### DISCUSSION THREAD

**Discussion by jrabary** [lancedb/lancedb#889] [General]
Hi all,
Lancedb is really a game changing for us and we would like to thank lancedb team for their effort.
In our tool for image data exploration and annotation https://github.com/pixano/pixano we heavily use lancedb in the backend. 
One use case that we are facing is adding new column to an existing table. Precisely, we have a table with object annotation with bounding box column in it. We want to annotate keypoints in a new version of the dataset. In this case we need to update the table schema. Usually, the type of the annotation are based on struct not a primitive type. This can somehow causes issue with pyarrow.
What is the best way to handle this if we want to keep it in lancedb level without going down to lance or pyarrow ?

**Comment by wjones127** [comment]
> One use case that we are facing is adding new column to an existing table.

Could you clarify, are you trying to add a new struct column, or are you trying to add a new column *within* an existing struct column? An example of the schema before and after would help clarify.

**Reply by tcoca27** [reply]
Hi @wjones127 
my use case involves adding a new column to an existing struct column.
so an example would be: 

Initial schema:
`class LanceSchema(LanceModel):
    id: str
    vector: Vector(1536)  # type: ignore
    metadata: LanceDocument

class LanceDocument(BaseModel):
    field1: Field1Type
    field2: Field2Type`

New Schema:
`class LanceSchema(LanceModel):
    id: str
    vector: Vector(1536)  # type: ignore
    metadata: LanceDocument

class LanceDocument(BaseModel):
   field1: Field1Type
   field2: Field2Type
   field3: Field3Type`

Is there a way to do this? Thanks

**Comment by jrabary** [comment]
let's say we have a table with the following initial schema:

```python
class InitialSchema(LanceModel):
   field1: Field1Type
   field2: Field2Type
```

Latter we want to add a new column to an existing table with type Field3type, so the final schema will become

```python
class FinalSchema(LanceModel):
   field1: Field1Type
   field2: Field2Type
   field3: Field3Type
```

How to migrate from the first table to the second one without loosing the data knowing that field3 maybe sparse. which means same row in the original table may have a not defined field3

**Reply by wjones127** [reply]
Thanks for clarifying. So there are two things here you need:

1. Need to add new columns to the dataset. You can currently do that with [LanceTable.merge()](https://github.com/lancedb/lancedb/blob/09cd08222d33ec391a23967c6244af7f30a2a501/python/lancedb/table.py#L962). We will be soon adding a new way to add columns in #896 (I'm actively working on this).
2. If `field3` is sparse, will need to add null values. Right now, LanceDB only supports null values for string columns. We are also actively working on this.
